File: fixtures/ext-types/lib/tests/bindings/test_imported_types.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import unittest
import urllib
from imported_types_lib import *
from imported_types_sublib import *
from uniffi_one_ns import *
from ext_types_custom import *

class TestIt(unittest.TestCase):
    def test_it(self):
        ct = get_combined_type(None)
        self.assertEqual(ct.uot.sval, "hello")
        self.assertEqual(ct.guid, "a-guid")
        self.assertEqual(ct.url.scheme, 'http')
        self.assertEqual(ct.url.netloc, 'example.com')
        self.assertEqual(ct.url.path, '/')

        ct2 = get_combined_type(ct)
        self.assertEqual(ct, ct2)

        t = get_trait_impl()
        self.assertEqual(t.hello(), "sub-lib trait impl says hello")
        sub = SubLibType(maybe_enum = None, maybe_trait = t, maybe_interface = None)
        self.assertTrue(get_sub_type(sub).maybe_trait is not None)

        ot = ObjectsType(maybe_trait = t, maybe_interface = None, sub = sub)
        self.assertTrue(ot.maybe_trait is not None)
        self.assertEqual(ot.maybe_interface, None)
        self.assertEqual(get_uniffi_one_trait(None), None)

        get_sub_type(sub)

    def test_get_url(self):
        url = urllib.parse.urlparse("http://example.com/")
        self.assertEqual(get_url(url), url)
        self.assertEqual(get_urls([url]), [url])
        self.assertEqual(get_maybe_url(url), url)
        self.assertEqual(get_maybe_url(None), None)
        self.assertEqual(get_maybe_urls([url, None]), [url, None])

    def test_custom_types(self):
        self.assertEqual(get_guid("guid"), "guid")
        self.assertEqual(get_ouid("ouid"), "ouid")
        self.assertEqual(get_ouid("uuid"), "uuid")
        self.assertEqual(get_nested_guid("uuid"), "uuid")

    def test_get_uniffi_one_type(self):
        t1 = UniffiOneType(sval="hello")
        self.assertEqual(t1, get_uniffi_one_type(t1))
        self.assertEqual(t1, get_maybe_uniffi_one_type(t1))
        self.assertEqual(None, get_maybe_uniffi_one_type(None))
        self.assertEqual([t1], get_uniffi_one_types([t1]))
        self.assertEqual([t1, None], get_maybe_uniffi_one_types([t1, None]))

    def test_get_uniffi_one_enum(self):
        e = UniffiOneEnum.ONE
        self.assertEqual(e, get_uniffi_one_enum(e))
        self.assertEqual(e, get_maybe_uniffi_one_enum(e))
        self.assertEqual(None, get_maybe_uniffi_one_enum(None))
        self.assertEqual([e], get_uniffi_one_enums([e]))
        self.assertEqual([e, None], get_maybe_uniffi_one_enums([e, None]))

    def test_external_crate_types(self):
        ct = get_combined_type(None)
        self.assertEqual(ct.ecd.sval, "ecd");
        self.assertEqual(get_external_crate_interface("foo").value(), "foo")

    def test_procmacro_types(self):
        t1 = UniffiOneProcMacroType(sval="hello")
        self.assertEqual(t1, get_uniffi_one_proc_macro_type(t1))
        self.assertEqual(t1, get_my_proc_macro_type(t1))

if __name__=='__main__':
    unittest.main()

File: fixtures/ext-types/lib/tests/bindings/test_imported_types.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import imported_types_lib
import Foundation

// First step: implement a trait from an external crate in Swift and pass it to a function from this
// crate.  This tests #2343 -- the codegen for this module needs to initialize the vtable from
// uniffi_one.
final class SwiftUniffiOneImpl: UniffiOneTrait {
    func hello() -> String {
        "Hello from Swift"
    }
}
assert(invokeUniffiOneTrait(t: SwiftUniffiOneImpl()) == "Hello from Swift")

let ct = getCombinedType(value: nil)
assert(ct.uot.sval == "hello")
assert(ct.guid ==  "a-guid")
assert(ct.url ==  URL(string: "http://example.com/"))

let ct2 = getCombinedType(value: ct)
assert(ct == ct2)

let t = getTraitImpl()
assert(t.hello() == "sub-lib trait impl says hello")
let sub = SubLibType(maybeEnum: nil, maybeTrait: t, maybeInterface: nil)
assert(getSubType(existing: sub).maybeTrait != nil)

let ob = ObjectsType(maybeTrait: t, maybeInterface: nil, sub: sub)
assert(getObjectsType(value: nil).maybeInterface == nil)
assert(getObjectsType(value: ob).maybeTrait != nil)
assert(getUniffiOneTrait(t: nil) == nil)

let url = URL(string: "http://example.com/")!;
assert(getUrl(url: url) ==  url)
assert(getMaybeUrl(url: url)! == url)
assert(getMaybeUrl(url: nil) == nil)
assert(getUrls(urls: [url]) == [url])
assert(getMaybeUrls(urls: [url, nil]) == [url, nil])

assert(getGuid(value: "guid") ==  "guid")
assert(getOuid(ouid: "ouid") ==  "ouid")
assert(getImportedOuid(ouid: "ouid") ==  "ouid")
assert(getNestedOuid(nouid: "ouid") ==  "ouid")
assert(getImportedNestedGuid(guid: nil) == "nested")
assert(getNestedExternalOuid(ouid: nil) == "nested-external-ouid")
assert(getImportedHandleU8(h: nil) == 3)

assert(getUniffiOneType(t: UniffiOneType(sval: "hello")).sval == "hello")
assert(getMaybeUniffiOneType(t: UniffiOneType(sval: "hello"))!.sval == "hello")
assert(getMaybeUniffiOneType(t: nil) == nil)
assert(getUniffiOneTypes(ts: [UniffiOneType(sval: "hello")]) == [UniffiOneType(sval: "hello")])
assert(getMaybeUniffiOneTypes(ts: [UniffiOneType(sval: "hello"), nil]) == [UniffiOneType(sval: "hello"), nil])
assert(getMyProcMacroType(t: UniffiOneProcMacroType(sval: "proc-macros")).sval == "proc-macros")

assert(getUniffiOneEnum(e: UniffiOneEnum.one) == UniffiOneEnum.one)
assert(getMaybeUniffiOneEnum(e: UniffiOneEnum.one)! == UniffiOneEnum.one)
assert(getMaybeUniffiOneEnum(e: nil) == nil)
assert(getUniffiOneEnums(es: [UniffiOneEnum.one]) == [UniffiOneEnum.one])
assert(getMaybeUniffiOneEnums(es: [UniffiOneEnum.one, nil]) == [UniffiOneEnum.one, nil])

do {
    try throwUniffiOneError()
    fatalError("Should have thrown")
} catch let e as UniffiOneError {
    if case let .Oops(reason) = e {
        assert(reason == "oh no")
    } else {
        fatalError("wrong error variant: \(e)")
    }
}

var counter = DispatchGroup()
counter.enter()
Task {
    do {
        try await throwUniffiOneErrorAsync()
        fatalError("Should have thrown")
    } catch let e as UniffiOneError {
        if case let .Oops(reason) = e {
            assert(reason == "oh no - async")
        } else {
            fatalError("wrong error variant: \(e)")
        }
    }
    counter.leave()
}
counter.wait()

do {
    try throwUniffiOneErrorInterface()
    fatalError("Should have thrown")
} catch let e as UniffiOneErrorInterface {
    assert(e.message() == "interface oops")
}

assert(ct.ecd.sval == "ecd")
assert(getExternalCrateInterface(val: "foo").value() == "foo")
